rs/ledger_suite/tests/sm-tests/src/lib.rs

```diff
@@ -2657,6 +2658,28 @@ pub fn test_upgrade_serialization(
                     test_upgrade(ledger_wasm_nextmigrationversionmemorymanager.clone());
                     // Test upgrade to memory manager again
                     test_upgrade(ledger_wasm_nextmigrationversionmemorymanager);
+
+                    // Current mainnet ICP wasm cannot deserialize from memory manager, but ICRC can
+                    match env.upgrade_canister(
+                        ledger_id,
+                        ledger_wasm_mainnet.clone(),
+                        upgrade_args.clone(),
+                    ) {
+                        Ok(_) => {
+                            if !downgrade_to_mainnet_should_succeed {
+                                panic!("Downgrade from memory manager directly to mainnet should fail (since mainnet is V0)!")
+                            }
+                        }
```

Comment:
now we're back at the mainnet version and will go _up_ to the current _again_ in the next step instead of _down_ to the current

Reply:
I reread the test again, and IIUC, the flow is as follows:

1. Install the ledger canister with the mainnet version `ledger_wasm_mainnet` (which is currently V0 for ICP, V1 for the ICRC ledger) (line 2618)
2. (Apply transactions)
3. Upgrade the ledger twice with `ledger_wasm_current`, which is V1 for both ledgers (lines 2651, 2653)
4. If specified, upgrade the ledger twice to `ledger_wasm_nextmigrationversionmemorymanager`, which is V2 for both ledgers (built with a feature flag used only for testing) (lines 2658, 2660)
5. Downgrade from V2 to the mainnet version `ledger_wasm_mainnet` (V0 for ICP, V1 for ICRC) (line 2663).
   1. In case of the ICP ledger, this downgrade shall fail, since downgrading from V2 to V0 is not supported (writing with the memory manager, reading only from raw stable memory)
   2. In case of the ICRC ledger, this downgrade succeeds, since it is now only downgrading from V2 to V1, and V1 can read data written using the memory manager.

Does this match your understanding? Are there changes needed?

Reply:
> In case of the ICRC ledger, this downgrade succeeds, since it is now only downgrading from V2 to V1,

this is based on the assumption that mainnet = current, right? in the general case, the test should do V0 -> V1 -> V1 (-> V2 -> V2 -> V1, if applicable) -> V0, but the code actually does: V0 -> V1 -> V1 -> V2 -> V2 -> V0 -> V1 -> V0 (meaning that V2 -> V1 is untested); if V0=V1 (as you seem to point out), then there's no difference, but in the general case, we seem to be lacking some coverage

Reply:
> this is based on the assumption that mainnet = current, right?

For the ICRC ledger this is now the case, yes!

Thanks for spelling it out with the example! The original test was doing V0 -> V1 -> V1 -> V2 -> V2 -> V0 (this would fail, so the ledger would still be at V2) -> V1 -> V0, which would cover all the cases. Now, since mainnet ICRC is at V1, we are in fact missing the V2 -> V1 for the ICRC ledger - I'll add this. In a separate PR, I will add fixed versions of the ledgers that are deployed on mainnet, so that if the versions currently in `mainnet-canisters.bzl` are updated, the test doesn't need to be adapted/no new test coverage gaps are introduced.

Reply:
> if the versions currently in mainnet-canisters.bzl are updated, the test doesn't need to be adapted/no new test coverage gaps are introduced

I don't see how this would make any difference

Reply:
Fair enough, on second thought maybe that doesn't make sense. The test should now go through all the migration paths for both ledgers though, PTAL.